Cache disable flag to avoid reading environmental variable.

fujitatomoya opened the pull request:
address https://github.com/ros2/rcl/issues/977

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

fujitatomoya commented:
@clalancette minor performance improvement. this can avoid reading environmental variable every time it publishes data with using `LoanedMessage`.

CC: @MichaelOrlov would be interested for you, as performance improvement.

fujitatomoya commented on rcl/include/rcl/publisher.h:
```diff
@@ -37,6 +37,8 @@ typedef struct rcl_publisher_s
 {
   /// Pointer to the publisher implementation
   rcl_publisher_impl_t * impl;
+  /// Disable flag to loan messages
+  bool disable_loaned_message;
```
Or maybe this can be in `rcl_publisher_option`? currently we do not have this option via rclcpp, so integrated in `rcl` layer with environmental variable would be good enough?

clalancette replied:
It would be nice to make this a publisher option, so we *could* control it via rclcpp if we wanted in the future.  I'd vote for moving it there.  Same goes for the subscription.

fujitatomoya replied:
okay, so we will do.

- extend option for publisher and subscription that can be set by user application via rclcpp.
- environmental variable check during initialization for publisher and subscription, and if environmental disable flag is on, that prevails option setting in the argument.

MichaelOrlov replied:
@fujitatomoya @clalancette I agree and it was my original proposal to add `disable_loaned_message` to the `rcl_publisher_options_t` and `rcl_subscription_options_t`.
Perhaps we can read out environment variable and init `disable_loaned_message` in `rcl_publisher_get_default_options()`
https://github.com/ros2/rcl/blob/d651a5cc77d3b81b1d779e611cae651238960206/rcl/src/rcl/publisher.c#L197-L207
and in `rcl_subscription_get_default_options()` https://github.com/ros2/rcl/blob/d651a5cc77d3b81b1d779e611cae651238960206/rcl/src/rcl/subscription.c#L200-L210
respectively.

fujitatomoya replied:
addressed by https://github.com/ros2/rcl/pull/1029/commits/3b25f4b8fa1043d8c83c490d2d6d6c41f1780132

clalancette commented on rcl/src/rcl/publisher.c:
```diff
@@ -104,6 +104,13 @@ rcl_publisher_init(
   RCL_CHECK_FOR_NULL_WITH_MSG(
     publisher->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; goto cleanup);
 
+  // Cache disable flag to loan messasges on this publisher
+  bool disable_loaned_message = false;
+  ret = rcl_get_disable_loaned_message(&disable_loaned_message);
+  if (ret == RCL_RET_OK && disable_loaned_message) {
+    publisher->disable_loaned_message = true;
+  }
```
I think we need to deal with errors better here:

```suggestion
  // Cache disable flag to loan messages on this publisher
  bool disable_loaned_message = false;
  ret = rcl_get_disable_loaned_message(&disable_loaned_message);
  if (ret != RCL_RET_OK) {
    goto cleanup;
  }
  publisher->disable_loaned_message = disable_loaned_message;
```

Same in the subscription below.

fujitatomoya replied:
thanks, this is a bug. i will fix it.

fujitatomoya commented:
@clalancette requesting 2nd review, modification for test code is bigger than fix itself...

fujitatomoya commented:
can we backport this humble? probably not, since c structure is extended, this is API compatibility but ABI?

fujitatomoya commented:
https://build.ros2.org/job/Rpr__rcl__ubuntu_jammy_amd64/154/ is related to this PR, but cannot reproduce this failure on my local environment.

@ros-pull-request-builder retest this please

fujitatomoya commented:
either @Barry-Xu-2018 or @iuhilnehc-ynos can you run the test with your local environment? I cannot reproduce the failure reported https://build.ros2.org/job/Rpr__rcl__ubuntu_jammy_amd64/159/.

```bash
colcon test --event-handlers console_direct+ --packages-select rcl
```

iuhilnehc-ynos commented:
@fujitatomoya 

I can reproduce this issue in the ros:rolling image with this PR. The failure in Rpc is because not using the latest rmw_fastrtps.
Please use CI directly.

fujitatomoya commented:
> It means the ros-rolling-rmw-fastrtps-cpp should be updated in http://repo.ros2.org/ubuntu/testing.

Ah, okay so using latest repo for rolling and CI should be fine. i will start the CI to verify. @iuhilnehc-ynos thanks for the information.

@clalancette i thought Rpr job base runtime is updated nightly, but obviously it is not. How often does it get updated?

clalancette commented:
> @clalancette i thought Rpr job base runtime is updated nightly, but obviously it is not. How often does it get updated?

Only when we do releases into Rolling.  We are way overdue for that; we will spend some time next week doing releases into Rolling.

fujitatomoya commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18021)](http://ci.ros2.org/job/ci_linux/18021/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12547)](http://ci.ros2.org/job/ci_linux-aarch64/12547/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18656)](http://ci.ros2.org/job/ci_windows/18656/)

fujitatomoya commented:
Window is unstable, i will restart it later.

fujitatomoya commented:
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18670)](http://ci.ros2.org/job/ci_windows/18670/)

MichaelOrlov commented on rcl/test/rcl/test_publisher.cpp:
```diff
@@ -372,38 +372,69 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_loan)
+      rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    });
+    if (is_fastdds) {
+      EXPECT_TRUE(rcl_publisher_can_loan_messages(&publisher));
+    } else {
+      EXPECT_FALSE(rcl_publisher_can_loan_messages(&publisher));
+    }
```
@fujitatomoya Could you please explain why there are different behavior for `rmw_fastrtps`?  
I think we shouldn't rely on default value of the environment variable in tests since it tend to be flaky or false negative/positive.
Instead we need explicitly set `publisher_options.disable_loaned_message` in code before publisher init to verify that our explicit options settings will override environment variable settings.

fujitatomoya replied:
@MichaelOrlov rmw_fastrtps just uses shared memory if possible as long as data type is POD, that is the default behavior for rmw_fastrtps, and this default behavior depends on rmw implementation. this default behavior verification was there before this PR, and i think we can keep those as well.

> explicitly set publisher_options.disable_loaned_message in code before publisher init to verify that our explicit options settings will override environment variable settings.

makes sense, i will add more test accordingly.

MichaelOrlov replied:
@fujitatomoya I had in mind that if you will explicitly set `publisher_options.disable_loaned_message = true` before calling publisher init the further call for the `rcl_publisher_can_loan_messages(&publisher)` should return always false - regardless of the internal `rmw_fastrtps` implementation.
This is actually what we need to check.

fujitatomoya replied:
addressed by https://github.com/ros2/rcl/pull/1029/commits/c6f9fc3e78c8afdf5fd2f1ad274fb7825e05f36f

MichaelOrlov reviewed (changes requested):
@fujitatomoya Thanks for moving to rcl_options and checking environment variable inside get_default_options methods. Implementation looks good to me.

However I have a question and suggestion for unit tests.
- Need to check that explicitly settled up value for the `publisher_options.disable_loaned_message` and `subscriber_options.disable_loaned_message` will not be overridden by the environment variable after calling pub/sub init.
- And we shouldn't rely on default value of the environment variable. We should set environment variable explicitly for each test.

fujitatomoya commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18028)](http://ci.ros2.org/job/ci_linux/18028/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12553)](http://ci.ros2.org/job/ci_linux-aarch64/12553/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18672)](http://ci.ros2.org/job/ci_windows/18672/)

fujitatomoya commented:
@iuhilnehc-ynos can you review this?

MichaelOrlov reviewed (changes requested):
@fujitatomoya Thanks for addressing comments from review.
I still have a strong opinion that we should not rely on default settings for `ROS_DISABLE_LOANED_MESSAGES` environment variable in `test_publisher_loan_disable` and `test_subscription_loan_disable` when doing check for `EXPECT_FALSE(publisher_options.disable_loaned_message);` right after reading default options.  
We need either explicitly set environment variable in test or even better remove check for `EXPECT_FALSE(publisher_options.disable_loaned_message);` after getting default options.  And always forcefully set `publisher_options.disable_loaned_message` value as we needed for test case.
The rationale for that is:
1. It's not safe to rely (not reliable to assume) that `ROS_DISABLE_LOANED_MESSAGES` environment variable will not be set by default on testing machine and environment. It could be leftovers from another tests. It might work currently but could be broken in future if order of tests will be changed or some new tests would be added.
1. You've already have dedicated tests for checking `publisher_options.disable_loaned_message` after getting default options with different values of the `ROS_DISABLE_LOANED_MESSAGES` environment variable.

iuhilnehc-ynos commented on rcl/include/rcl/publisher.h:
```diff
@@ -49,6 +49,8 @@ typedef struct rcl_publisher_options_s
   rcl_allocator_t allocator;
   /// rmw specific publisher options, e.g. the rmw implementation specific payload.
   rmw_publisher_options_t rmw_publisher_options;
+  /// Disable flag to LoanedMessage, initialized via environmental variable.
```
To make the comment clear, do you think it's necessary to add something like "True only if `ROS_DISABLE_LOANED_MESSAGES=1`"? (some cases might be used, e.g:  ROS_DISABLE_LOANED_MESSAGES=TRUE)

https://github.com/ros2/rcl/blob/c6f9fc3e78c8afdf5fd2f1ad274fb7825e05f36f/rcl/include/rcl/publisher.h#L198 should be updated

```
 * - rmw_publisher_options = rmw_get_default_publisher_options()
 * - disable_loaned_message = ?
```
Sorry, I have no good description for this.

iuhilnehc-ynos commented on rcl/src/rcl/publisher.c:
```diff
@@ -203,6 +203,16 @@ rcl_publisher_get_default_options()
+  if (ret == RCL_RET_OK) {
+    default_options.disable_loaned_message = disable_loaned_message;
+  } else {
+    default_options.disable_loaned_message = false;
```
To call `rcl_reset_error` in `else block` might be better because `rcl_get_disable_loaned_message` sets the error message when returns ERROR, right?

What about
```
  bool disable_loaned_message = false;
  rcl_ret_t ret = rcl_get_disable_loaned_message(&disable_loaned_message);
  if (ret != RCL_RET_OK) {
    // debug log
    rcl_reset_error();
  }
  default_options.disable_loaned_message = disable_loaned_message;
```

fujitatomoya replied:
yeah, right. i will fix this as you suggested.

fujitatomoya replied:
actually i just wanted to use local value only when API succeeds, but i will take some of your suggestion here.

iuhilnehc-ynos reviewed (commented):
LGTM with some minor comments.

fujitatomoya commented:
> @fujitatomoya Thanks for addressing comments from review. I still have a strong opinion that we should not rely on default settings for `ROS_DISABLE_LOANED_MESSAGES` environment variable in `test_publisher_loan_disable` and `test_subscription_loan_disable` when doing check for `EXPECT_FALSE(publisher_options.disable_loaned_message);` right after reading default options. We need either explicitly set environment variable in test or even better remove check for `EXPECT_FALSE(publisher_options.disable_loaned_message);` after getting default options. And always forcefully set `publisher_options.disable_loaned_message` value as we needed for test case. The rationale for that is:
> 
> 1. It's not safe to rely (not reliable to assume) that `ROS_DISABLE_LOANED_MESSAGES` environment variable will not be set by default on testing machine and environment. It could be leftovers from another tests. It might work currently but could be broken in future if order of tests will be changed or some new tests would be added.
> 2. You've already have dedicated tests for checking `publisher_options.disable_loaned_message` after getting default options with different values of the `ROS_DISABLE_LOANED_MESSAGES` environment variable.

@MichaelOrlov I see, probably it would be better, not to have flaky test 👍

fujitatomoya commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18029)](http://ci.ros2.org/job/ci_linux/18029/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12554)](http://ci.ros2.org/job/ci_linux-aarch64/12554/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18673)](http://ci.ros2.org/job/ci_windows/18673/)

fujitatomoya commented:
@MichaelOrlov @iuhilnehc-ynos thanks for the review, all comments are addressed. requesting final check.

iuhilnehc-ynos commented on rcl/src/rcl/publisher.c:
```diff
@@ -203,6 +203,19 @@ rcl_publisher_get_default_options()
+    default_options.disable_loaned_message = disable_loaned_message;
+  } else {
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Failed to get disable_loaned_message: ");
+    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
```
I will use `RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING` instead, but it's OK.

fujitatomoya commented:
@iuhilnehc-ynos thanks for the review.

CI(windows unstable):
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18676)](http://ci.ros2.org/job/ci_windows/18676/)

MichaelOrlov commented on rcl/src/rcl/subscription.c:
```diff
@@ -213,6 +213,9 @@ rcl_subscription_get_default_options()
   } else {
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Failed to get disable_loaned_message: ");
+    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
+    rcl_reset_error();
```
The same comment in regards removing logging to the error stream

fujitatomoya replied:
>  And by default it will be absent in most configurations.

I think that this error is important cz using `ROS_DISABLE_LOANED_MESSAGES` is meant to disable the feature. so this is strong user intention. in that case if unexpected error happens, it should be able to notify the user that `ROS_DISABLE_LOANED_MESSAGES` is not effectively working.

> This error string will appear each time we creating new pub/sub - it would be pretty annoying and could also affect performance.

I believe there should be. I think it is unlikely that this logging affects the performance, because it only happens in initialization. I think originally this PR is targeting send/receive performance improvement but initialization or finalization?

what do you think?

MichaelOrlov replied:
@fujitatomoya Ahh never mind. 
rcl_get_disable_loaned_message(bool * disable_loaned_message) can return `RCL_RET_ERROR` only if input parameters are wrong which is not a our case any way.  
You can keep your output to the error string but error will never happened.
I was wrong on assumption that `rcl_get_disable_loaned_message(bool * disable_loaned_message)` that will return some error code if `ROS_DISABLE_LOANED_MESSAGES` env variable not defined. Sorry for confusion.

fujitatomoya replied:
> You can keep your output to the error stream but error will never happened.

yeah you are right, this error might print here if `something` really unexpected happens. probably never... 😅 

not at all, really appreciate your review and comments, thank you so much 👍

MichaelOrlov reviewed (changes requested):
@fujitatomoya Please remove logging to the stderr stream when rcl_get_disable_loaned_message(&disable_loaned_message) returning failure.
This is a normal situation when `ROS_DISABLE_LOANED_MESSAGES` is not defined and `ret != RCL_RET_OK` in this particular situation.

fujitatomoya commented:
@clalancette i am putting a hold on this until https://github.com/ros2/rcl/pull/1029#issuecomment-1414028600 is resolved. after rolling release, i will kick the Rpr job to make sure it passes, and then i will merge this into rolling.

fujitatomoya commented:
@ros-pull-request-builder retest this please

fujitatomoya commented:
rpr job is happy now, CI is green.

@clalancette i will be waiting for the final review on this.

fujitatomoya commented:
@clalancette @MichaelOrlov @iuhilnehc-ynos thank you for the review, i will go ahead to merge this. since this changes ABI, this will be available on Iron or later.
